gspheres/vish.py

```diff
@@ -0,0 +1,229 @@
+from typing import Union
+import numpy as np
+import tensorflow as tf
+import gpflow
+from gpflow.inducing_variables import InducingVariables
+from gpflow import covariances as cov
+from gpflow import kullback_leiblers as kl
+from gpflow.base import TensorLike
+from gpflow.models import SVGP
+from gpflow.models.model import MeanAndVariance
+from gpflow.models.training_mixins import InputData
+from gpflow.utilities import to_default_float
+
+from gspheres.kernels.spherical_matern import SphericalMatern
+from gspheres.spherical_harmonics import SphericalHarmonics
+
+
+def map_to_sphere(X: np.ndarray, bias: Union[int, float]) -> np.ndarray:
+    """
+    Append bias to X and map to surface of the unit hypersphere.
+
+    X is [N, D]
+    """
+    Xb = np.concatenate((X, bias * np.ones((*X.shape[:-1], 1))), axis=1)
+    return Xb / np.linalg.norm(Xb, axis=1, keepdims=True)
+
+
+# Define InducingVariables
+class SphericalHarmonicFeatures(InducingVariables):
+    """Wraps SphericalHarmonics."""
+    def __init__(self, dimension, degrees):
+        self.spherical_harmonics = SphericalHarmonics(dimension, degrees)
+
+    def __len__(self):
+        """Number of inducing variables"""
+        return len(self.spherical_harmonics)
+
+
+@cov.Kuu.register(SphericalHarmonicFeatures, SphericalMatern)
+def Kuu_sphericalmatern_sphericalharmonicfeatures(
+        inducing_variable,
+        kernel,
+        jitter=None
+):
+    """Covariance matrix between spherical harmonic features."""
+    eigenvalues = kernel.eigenvalues(len(inducing_variable))
```

Comment:
This doesn't look exactly right. `kernel.eigenvalues(n)` returns the eigenvalues `\lambda_n`'s for the first n levels. However, for each level, we have `N(d, n)` number of spherical harmonics. We need to make sure to have the correct repetition of eigenvalues. I've done this before using the "chain" function: https://github.com/vdutor/GPsOnSpheres/blob/master/scripts/plot_samples.py#L14.

Reply:
```suggestion
    eigenvalues_per_level = kernel.eigenvalues(inducing_variable.max_degree)
    num_harmonics_per_level = tf.convert_to_tensor(
        [num_harmonics(dimension, n) for n in range(inducing_variable.max_degree)]
    )
    eigenvalues = chain(eigenvalues_per_level, num_harmonics_per_level)
```

Okay, cheers for pointing that out, I think that your code from plot_samples can just be copied over, and we'll add max_degree as an attribute of inducing_variable upon instantiation.

Reply:
Nice, feel free to directly commit changes to this branch.